config/config_test.go

```diff
@@ -107,6 +108,123 @@ func TestSampleConfig(t *testing.T) {
+			inConfigFileName: "non_existent_file",
+			out: testOut{
+				expectedLogInfo: []logComponents{
+					{
+						msg: "Failed to load config file:",
+						lvl: logrus.FatalLevel,
+					},
+				},
+			},
+		},
+		{
+			description:      "Non-yaml configuration file exists and is valid: stop execution and log Fatal message",
+			inConfigFileName: filepath.Join("configtest", "json_config"),
+			out: testOut{
+				expectedLogInfo: []logComponents{
+					{
+						msg: "Failed to load config file:",
+						lvl: logrus.FatalLevel,
+					},
+				},
+			},
+		},
+		{
+			description:      "file exists but its yaml markup is invalid: stop execution and log Fatal message",
+			inConfigFileName: filepath.Join("configtest", "config_invalid"),
+			out: testOut{
+				expectedLogInfo: []logComponents{
+					{
+						msg: "Failed to load config file:",
+						lvl: logrus.FatalLevel,
+					},
+				},
+			},
+		},
+		{
+			description:      "Valid yaml file exists, configuration from file gets read but does not override port default value",
+			inConfigFileName: filepath.Join("configtest", "config"),
+			out:              testOut{},
+		},
+		{
+			description:      "Valid yaml configuration gets parsed from file and overrides port default value",
+			inConfigFileName: filepath.Join("configtest", "config_overrides_defaults"),
+			out:              testOut{overridesDefaultPort: true},
+		},
+	}
+
+	//substitute logger exit function so execution doesn't get interrupted
+	defer func() { logrus.StandardLogger().ExitFunc = nil }()
+	logrus.StandardLogger().ExitFunc = func(int) {}
+
+	for _, tc := range testCases {
+		//run test
+		cfg := NewConfig(tc.inConfigFileName)
+
+		// Assert logrus expected entries
+		if assert.Len(t, hook.Entries, len(tc.out.expectedLogInfo), "Incorrect number of entries were logged to logrus in test: %s. Expected: %d Actual: %d", tc.description, len(tc.out.expectedLogInfo), len(hook.Entries)) {
+			for i := 0; i < len(tc.out.expectedLogInfo); i++ {
+				//assert.Equal(t, tc.out.expectedLogInfo[i].msg, hook.Entries[i].Message, "Wrong log message. Test: %s", tc.description)
+				assert.True(t, strings.HasPrefix(hook.Entries[i].Message, tc.out.expectedLogInfo[i].msg), "Wrong log message. Test: %s", tc.description)
+				assert.Equal(t, tc.out.expectedLogInfo[i].lvl, hook.Entries[i].Level, "Wrong info level. Test: %s", tc.description)
+			}
+		} else {
+			return
+		}
+
+		// Assert configuration
+		assert.Equal(t, defaultAdminPort, cfg.AdminPort, "AdminPort number in this test is supposed to be the 2525 default. Test: %s", tc.description)
+		if tc.out.overridesDefaultPort {
+			assert.NotEqual(t, defaultPort, cfg.Port, "Port number in this test is supposed to be different from the 2424 default. Test: %s", tc.description)
+		} else {
+			assert.Equal(t, 2424, cfg.Port, "Port number in this test is supposed to be 2424 default. Test: %s", tc.description)
+		}
+
+		//Reset log after every test and assert successful reset
+		hook.Reset()
```

Comment:
Please move to line 114 as a defer func, in case the test run never makes it this far.

Reply:
We need to `hook.Reset()` at the end of every loop and the `defer` keyword makes the call just before the surrounding function ends. In other words, `defer` would  make all the `hook.Reset()` calls once we exit the for loop just before exiting `TestNewConfigFuncFileParam`.

According to [this source](https://golangbot.com/defer) the following example prints the `String` backwards.
```go
func main() {  
    str := "String"
    fmt.Printf("Original String: %s\n", string(str))
    fmt.Printf("Reversed String: ")
    for _, v := range []rune(str) {
        defer fmt.Printf("%c", v)
    }
}
```

Reply:
👍 `defer` is LIFO. Can't we just move this snippet to the beginning of the loop?
```
hook.Reset()
assert.Nil(t, hook.LastEntry())
```

Reply:
This sentence that I wrote above is wrong: "In other words, `defer` would make all the `hook.Reset()` calls once we exit the for loop just before exiting `TestNewConfigFuncFileParam`."

I'm not sure what point I was trying to make but the thing is:`logrus.StandardLogger().ExitFunc = func(int) {}` is set so our test does not interrupt exection every time we log something to `FatalLevel`. `hook.Reset()` on
the other hand, gets called at the end of each loop iteration to clear the entries of the log. I moved the call to the beginning of the loop as you suggested and tests still pass maybe because the `hook` object goes
out of scope and gets garbage collected at the end of `TestNewConfigFuncFileParam`.

Even though the output doesn't change, do you think we should leave it at the end of the `for` loop?

Reply:
👍 I think it's ok to put it at the beginning or the end so I vote to leave it where you have it.